File: javascript/bestSum.js
function bestSum(targetSum, numbers) {
    if (targetSum === 0) return []
    if (targetSum < 0) return null

    let shortestCombination = null

    for (let num of numbers) {
        const remainderResult = bestSum(targetSum - num, numbers)
        const currentCombination = remainderResult ?  [...remainderResult, num] : null
        
        if (shortestCombination === null
            || currentCombination
            && currentCombination.length < shortestCombination.length)
            shortestCombination = currentCombination
    }

    return shortestCombination
}

console.log(bestSum(80, [5, 6, 3]))

File: kotlin/bestSum.kt
fun bestSum(targetSum: Int, numbers: List<Int>): List<Int>? {
    if (targetSum == 0) return emptyList()
    if (targetSum < 0) return null

    var shortestCombination: List<Int>? = null

    numbers.forEach() {
        val remainderResult: List<Int>? = bestSum(targetSum - it, numbers)
        val currentCombination: List<Int>? = remainderResult?.plus(it)

        if (shortestCombination == null
            || currentCombination != null
            && currentCombination.size < (shortestCombination as List).size)
            shortestCombination = currentCombination        
    }

    return shortestCombination
}

fun main(vararg args: String): Unit {
    println(bestSum(args[0].toInt(), mutableListOf<Int>(4, 8, 5)))
}